backup: TestBackupRestoreWithConcurrentWrites failed

backup.TestBackupRestoreWithConcurrentWrites [failed](https://mesolite.cluster.engflow.com/invocations/default/002ea988-35a7-4945-94a0-ebafdd7676df?testReportRun=1&testReportShard=47&testReportAttempt=1#targets-Ly9wa2cvYmFja3VwOmJhY2t1cF90ZXN0) on release-25.1 @ [52e0d07d04d4b8be2dd926bfd1257dbfde0869e9](https://github.com/cockroachdb/cockroach/commits/52e0d07d04d4b8be2dd926bfd1257dbfde0869e9):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBackupRestoreWithConcurrentWrites (15m0s)

```
Stack: 

```
goroutine 253179 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBackupRestoreWithConcurrentWrites
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestBackupRestoreWithConcurrentWrites3608492609
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=47</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139045 backup: TestBackupRestoreWithConcurrentWrites failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-master]
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreWithConcurrentWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-46529

backup.TestBackupRestoreWithConcurrentWrites [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18518548?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18518548?buildTab=artifacts#/) on release-25.1 @ [dc462b9ae9f5e135c1cee3cb479d6a4d4bd873cc](https://github.com/cockroachdb/cockroach/commits/dc462b9ae9f5e135c1cee3cb479d6a4d4bd873cc):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBackupRestoreWithConcurrentWrites (15m0s)

```
Stack: 

```
goroutine 373823 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBackupRestoreWithConcurrentWrites
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/6d8f0d953bcf81a6c6476724cd8eb3b1/logTestBackupRestoreWithConcurrentWrites562375861
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139045 backup: TestBackupRestoreWithConcurrentWrites failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreWithConcurrentWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestBackupRestoreWithConcurrentWrites [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18520408?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18520408?buildTab=artifacts#/) on release-25.1 @ [0518bc00124b6f9f5351fdfa5fe0b85af345ec17](https://github.com/cockroachdb/cockroach/commits/0518bc00124b6f9f5351fdfa5fe0b85af345ec17):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBackupRestoreWithConcurrentWrites (15m0s)

```
Stack: 

```
goroutine 449118 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBackupRestoreWithConcurrentWrites
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/6d8f0d953bcf81a6c6476724cd8eb3b1/logTestBackupRestoreWithConcurrentWrites1336446169
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139045 backup: TestBackupRestoreWithConcurrentWrites failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreWithConcurrentWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestBackupRestoreWithConcurrentWrites [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18529210?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18529210?buildTab=artifacts#/) on release-25.1 @ [ef9b4358edaa3b626106baeacbba2c174e82f274](https://github.com/cockroachdb/cockroach/commits/ef9b4358edaa3b626106baeacbba2c174e82f274):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBackupRestoreWithConcurrentWrites (15m0s)

```
Stack: 

```
goroutine 431648 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBackupRestoreWithConcurrentWrites
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/6d8f0d953bcf81a6c6476724cd8eb3b1/logTestBackupRestoreWithConcurrentWrites62192889
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139045 backup: TestBackupRestoreWithConcurrentWrites failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreWithConcurrentWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestBackupRestoreWithConcurrentWrites [failed](https://tanzanite.cluster.engflow.com/invocations/default/0f087c08-b16e-4281-a562-6c96ffb70d22?testReportRun=5&testReportShard=47&testReportAttempt=1#targets-Ly9wa2cvYmFja3VwOmJhY2t1cF90ZXN0) on release-25.1 @ [f9c2a2f417a567a1dc6e7b23b41b8759cf62d077](https://github.com/cockroachdb/cockroach/commits/f9c2a2f417a567a1dc6e7b23b41b8759cf62d077):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBackupRestoreWithConcurrentWrites (15m0s)

```
Stack: 

```
goroutine 327379 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBackupRestoreWithConcurrentWrites
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestBackupRestoreWithConcurrentWrites3532898514
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=5</code>
 - <code>shard=47</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139045 backup: TestBackupRestoreWithConcurrentWrites failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreWithConcurrentWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestBackupRestoreWithConcurrentWrites [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18539365?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18539365?buildTab=artifacts#/) on release-25.1 @ [f0f424368acbf7d62ea7a55d8dae564fa04f9059](https://github.com/cockroachdb/cockroach/commits/f0f424368acbf7d62ea7a55d8dae564fa04f9059):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBackupRestoreWithConcurrentWrites (15m0s)

```
Stack: 

```
goroutine 730760 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBackupRestoreWithConcurrentWrites
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/6d8f0d953bcf81a6c6476724cd8eb3b1/logTestBackupRestoreWithConcurrentWrites1050884570
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:152: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139045 backup: TestBackupRestoreWithConcurrentWrites failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreWithConcurrentWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestBackupRestoreWithConcurrentWrites [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18539241?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18539241?buildTab=artifacts#/) on release-25.1 @ [6cb042e04510e3838b1ae371ea915154de91d648](https://github.com/cockroachdb/cockroach/commits/6cb042e04510e3838b1ae371ea915154de91d648):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBackupRestoreWithConcurrentWrites (15m0s)

```
Stack: 

```
goroutine 420096 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBackupRestoreWithConcurrentWrites
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/6d8f0d953bcf81a6c6476724cd8eb3b1/logTestBackupRestoreWithConcurrentWrites1156494629
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139045 backup: TestBackupRestoreWithConcurrentWrites failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreWithConcurrentWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestBackupRestoreWithConcurrentWrites [failed](https://tanzanite.cluster.engflow.com/invocations/default/6d906e79-9f1c-4338-92cc-c8266d86a8c3?testReportRun=8&testReportShard=47&testReportAttempt=1#targets-Ly9wa2cvYmFja3VwOmJhY2t1cF90ZXN0) on release-25.1 @ [29c0a3d5161b6818e414efc9c35321c1c0416e19](https://github.com/cockroachdb/cockroach/commits/29c0a3d5161b6818e414efc9c35321c1c0416e19):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBackupRestoreWithConcurrentWrites (15m0s)

```
Stack: 

```
goroutine 745358 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBackupRestoreWithConcurrentWrites
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestBackupRestoreWithConcurrentWrites3675806789
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:152: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=8</code>
 - <code>shard=47</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139045 backup: TestBackupRestoreWithConcurrentWrites failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreWithConcurrentWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestBackupRestoreWithConcurrentWrites [failed](https://tanzanite.cluster.engflow.com/invocations/default/6a723732-b1cb-4f15-85f0-4a6c68bc048d?testReportRun=3&testReportShard=47&testReportAttempt=1#targets-Ly9wa2cvYmFja3VwOmJhY2t1cF90ZXN0) on release-25.1 @ [6123ebaa10c3847899aa81b09ec4cb8d08c0bd6d](https://github.com/cockroachdb/cockroach/commits/6123ebaa10c3847899aa81b09ec4cb8d08c0bd6d):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBackupRestoreWithConcurrentWrites (15m0s)

```
Stack: 

```
goroutine 297975 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBackupRestoreWithConcurrentWrites
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestBackupRestoreWithConcurrentWrites1940619797
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=3</code>
 - <code>shard=47</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139045 backup: TestBackupRestoreWithConcurrentWrites failed [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreWithConcurrentWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestBackupRestoreWithConcurrentWrites [failed](https://tanzanite.cluster.engflow.com/invocations/default/5f767eae-57b7-4be9-9575-418e8ebd2dfe?testReportRun=6&testReportShard=47&testReportAttempt=1#targets-Ly9wa2cvYmFja3VwOmJhY2t1cF90ZXN0) on release-25.1 @ [0a14af43960ce2fcb79a2624d3b94fe2ce3662e2](https://github.com/cockroachdb/cockroach/commits/0a14af43960ce2fcb79a2624d3b94fe2ce3662e2):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBackupRestoreWithConcurrentWrites (15m0s)

```
Stack: 

```
goroutine 399452 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBackupRestoreWithConcurrentWrites
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestBackupRestoreWithConcurrentWrites3099848663
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=47</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139045 backup: TestBackupRestoreWithConcurrentWrites failed [A-disaster-recovery C-test-failure O-robot P-2 T-sql-foundations branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreWithConcurrentWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

backup.TestBackupRestoreWithConcurrentWrites [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18578325?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18578325?buildTab=artifacts#/) on release-25.1 @ [0eade67d677676921027deed791fb1c3d83c0fc9](https://github.com/cockroachdb/cockroach/commits/0eade67d677676921027deed791fb1c3d83c0fc9):

Fatal error:

```
panic: test timed out after 15m0s
running tests:
	TestBackupRestoreWithConcurrentWrites (15m0s)

```
Stack: 

```
goroutine 381724 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x30c
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBackupRestoreWithConcurrentWrites
I250122 22:50:14.403161 1 (gostd) rand.go:246  [-] 1  random seed: -9023316714843328118
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/6d8f0d953bcf81a6c6476724cd8eb3b1/logTestBackupRestoreWithConcurrentWrites147109015
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #139045 backup: TestBackupRestoreWithConcurrentWrites failed [A-disaster-recovery C-test-failure O-robot P-2 T-sql-foundations branch-master]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreWithConcurrentWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/catalog/lease/descriptor_state.go
+++ b/pkg/sql/catalog/lease/descriptor_state.go
@@ -315,10 +315,15 @@ func (t *descriptorState) release(ctx context.Context, s *descriptorVersionState
 }
 
 // maybeQueueLeaseRenewal queues a lease renewal if there is not already a lease
-// renewal in progress.
+// renewal in progress. On error the leased descriptor will be released.
 func (t *descriptorState) maybeQueueLeaseRenewal(
-	ctx context.Context, m *Manager, id descpb.ID, name string,
-) error {
+	ctx context.Context, m *Manager, ld LeasedDescriptor,
+) (retError error) {
+	defer func() {
+		if retError != nil {
+			ld.Release(ctx)
+		}
+	}()
 	if !atomic.CompareAndSwapInt32(&t.renewalInProgress, 0, 1) {
 		return nil
 	}
@@ -330,7 +335,7 @@ func (t *descriptorState) maybeQueueLeaseRenewal(
 	newCtx = logtags.AddTags(newCtx, logtags.FromContext(ctx))
 	return t.stopper.RunAsyncTask(newCtx,
 		"lease renewal", func(ctx context.Context) {
-			t.startLeaseRenewal(ctx, m, id, name)
+			t.startLeaseRenewal(ctx, m, ld.GetID(), ld.GetName())
 		})
 }
 

--- a/pkg/sql/catalog/lease/lease.go
+++ b/pkg/sql/catalog/lease/lease.go
@@ -1345,7 +1345,7 @@ func (m *Manager) AcquireByName(
 			if durationUntilExpiry < m.storage.leaseRenewalTimeout() {
 				if t := m.findDescriptorState(descVersion.GetID(), false /* create */); t != nil {
 					if err := t.maybeQueueLeaseRenewal(
-						ctx, m, descVersion.GetID(), name); err != nil {
+						ctx, m, descVersion); err != nil {
 						return nil, err
 					}
 				}
@@ -1512,7 +1512,7 @@ func (m *Manager) Acquire(
 			if latest {
 				durationUntilExpiry := time.Duration(desc.getExpiration(ctx).WallTime - timestamp.WallTime)
 				if durationUntilExpiry < m.storage.leaseRenewalTimeout() {
-					if err := t.maybeQueueLeaseRenewal(ctx, m, id, desc.GetName()); err != nil {
+					if err := t.maybeQueueLeaseRenewal(ctx, m, desc); err != nil {
 						return nil, err
 					}
 				}
@@ -1694,11 +1694,14 @@ func (m *Manager) RefreshLeases(ctx context.Context, s *stop.Stopper, db *kv.DB)
 						if err != nil {
 							log.Warningf(ctx, "error fetching lease for descriptor %s", err)
 						}
-					} else {
-						if err := purgeOldVersions(ctx, db, desc.GetID(), dropped, desc.GetVersion(), m); err != nil {
-							log.Warningf(ctx, "error purging leases for descriptor %d(%s): %s",
-								desc.GetID(), desc.GetName(), err)
-						}
+					}
+					// Even if an initial acquisition happens above, we need to purge old
+					// descriptor versions, which could have been acquired concurrently.
+					// For example the range feed sees version 2 and a query concurrently
+					// acquires version 1.
+					if err := purgeOldVersions(ctx, db, desc.GetID(), dropped, desc.GetVersion(), m); err != nil {
+						log.Warningf(ctx, "error purging leases for descriptor %d(%s): %s",
+							desc.GetID(), desc.GetName(), err)
 					}
 				}
 				// New descriptors may appear in the future if the descriptor table is

